src/loader/dag/const_collapse.rs

```diff
@@ -0,0 +1,165 @@
+                    finder
+                        .find_constant_inputs(&node.inputs, prev_nodes, const_block_inputs.clone())
+                        .1
+                        .iter()
+                        .enumerate()
+                        .filter_map(|(idx, maybe_const)| match maybe_const {
+                            MaybeConstant::ReferenceConstant { value, .. } => {
+                                Some((idx as u32, value.clone()))
+                            }
+                            MaybeConstant::CollapsedConstant(value) => {
+                                Some((idx as u32, value.clone()))
+                            }
+                            MaybeConstant::NonConstant => None,
+                        })
+                        .collect()
+                } else {
+                    // Loops have multiple entry points, so we can't be sure a constant input
+                    // at entry will be always constant. Besides, an optimized WASM wouldn't
+                    // have constant inputs to loops anyway.
+                    HashMap::new()
+                };
+
+                num_collapsed += recursive_constant_collapse(
+                    finder,
+                    processor,
+                    &mut sub_dag.nodes,
+                    const_inputs,
+                );
+            }
+            _ => {
+                // Constant collapse is not supported for other node types.
+            }
+        }
+    }
+
+    num_collapsed
+}
+
+struct ConstantInputFinder {
+    /// This buffer will be reused to avoid allocations on each node.
+    buffer: Vec<MaybeConstant>,
+}
+
+impl ConstantInputFinder {
+    fn new() -> Self {
+        Self { buffer: Vec::new() }
+    }
+
+    /// Finds which of the given inputs are constants.
+    fn find_constant_inputs<'a>(
+        &'a mut self,
+        inputs: &[NodeInput],
+        nodes: &mut [dag::Node],
+        const_block_inputs: HashMap<u32, WasmValue>,
+    ) -> (bool, &'a [MaybeConstant]) {
+        let mut has_constant = false;
+
+        // Check if any of the inputs are constants.
+        self.buffer.clear();
+        self.buffer.extend(inputs.iter().map(|input| {
+            let origin = match input {
+                NodeInput::Reference(value_origin) => value_origin,
+                NodeInput::Constant(wasm_value) => {
+                    // Weird... there shouldn't be any collapsed constants before this pass.
+                    // But we can handle it anyway.
+                    return MaybeConstant::CollapsedConstant(wasm_value.clone());
+                }
+            };
+
+            match &nodes[origin.node].operation {
+                Operation::Inputs => {
+                    // This refers to the input of the block, we need to check if it's a constant.
+                    const_block_inputs.get(&origin.output_idx).cloned()
+                }
+                Operation::WASMOp(op) => op.try_into().ok(),
```

Comment:
can you add a small comment here, saying its converting "const producing" operations to the resulting value?
idk why but this match took me a while to parse even though its really simple, maybe because of the `try_into` idk...

Reply:
fixed